### PR TITLE
rec: implement catalog zones to populate forwarding zones

### PR DESCRIPTION
**Pull request by omoerbeek**
### Short description

Supersedes  #12865

Example of a config:
```yaml
recursor:
  forwarding_catalog_zones:
  - zone: catalog.example
    notify_allowed: true
    xfr:
      addresses: [127.0.0.1]
    groups:
    - forwarders: [1.2.3.4, '[::123]:99']
    - name: mygroup
       forwarders: [4.5.6.7]
       recurse: true
       notify_allowed: true
```

See docs for more explanation.
Includes a refactoring of some of the existing XFR code, but not fully. Doing a full refactor of the rpzloader code is for another time.

### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [X] compiled this code
- [X] tested this code
- [X] included documentation (including possible behaviour changes)
- [X] documented the code
- [X] added or modified regression test(s)
- [X] added or modified unit test(s)

### CODE REVIEW

File: pdns/recursordist/settings/docs-new-preamble-in.rst [hunk 1]
```diff
@@ -472,6 +472,57 @@ An example of an ``proxymappings`` entry, which is a sequence of `ProxyMapping`_
...
+        forwarders: [192.168.179.2] # forwarder for catalog zone members in mygroup
+        recurse: true
+        notify_allowed: true
+   - zone: 'forward2.example'
+     xfr:
+       addresses: [128.66.1.3]
+     groups:
+      - forwarders: [192.168.178.3] # only default forwarder for 2nd catalog zone
+
+:program:`Recursor` will transfer the catalog zone from te authoritative server using IXFR (falling back to AXFR if needed) and add forwarding clauses for all members of the catalog zone.
```
Comment (phonedph1) [inline pdns/recursordist/settings/docs-new-preamble-in.rst:520]: typo here "te"

Comment (jsoref) [reply pdns/recursordist/settings/docs-new-preamble-in.rst:520]: if you think `the` is _likely_ to be misspelled as `te`, then you could add this to `line_forbidden.patterns`:
```
# Should probably be `the`
\bte\b
```

**Comment by omoerbeek** [comment]
Rebased to fix conflict.

File: pdns/recursordist/rec-xfr.hh [hunk 2]
```diff
@@ -47,27 +52,97 @@ struct ZoneXFRParams
...
+  }
+  void setName(const DNSName& name)
+  {
+    d_name = name;
+  }
+  [[nodiscard]] auto getName() const
+  {
+    return d_name;
+  }
+  void reserve([[maybe_unused]] size_t size)
```
[the remark this answers is not in the document]
Comment (omoerbeek) [reply pdns/recursordist/rec-xfr.hh:79]: These are remains af an attempt to merge the RPZ zone transfer and the new catz transfer. For example `reserve` can be implemented in the latter, but not in the former as it used a different data structure.  I never managed to finish that, so they remain separate. But see the remains are confusing, will remove.

File: pdns/recursordist/rec-xfr.cc [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+    }
+    if (!values.emplace(ptr->getContent()).second) {
+      invalid = true;
+      break;
+    }
+  }
+  return !invalid;
+}
+
+static shared_ptr<const SOARecordContent> loadZoneFromServer(Logr::log_t plogger, const ComboAddress& primary, const DNSName& zoneName, shared_ptr<CatalogZone>& zone, const TSIGTriplet& tsigTriplet, size_t maxReceivedBytes, const ComboAddress& localAddress, uint16_t axfrTimeout)
```
Comment (rgacogne) [inline pdns/recursordist/rec-xfr.cc:174]: Not an issue but the name and the first log lines feel like this function is intended to load generic zones, not only catalog zones, then by the end we have catalog-specific processing which surprised me (but is fine).

File: pdns/recursordist/rec-xfr.hh [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+public:
+  FWCatZoneXFR(ZoneXFRParams params) :
+    ZoneXFR(std::move(params))
+  {}
+
+  static void zoneXFRTracker(ZoneXFRParams params, uint64_t configGeneration);
+
+private:
+  void preloadZoneFile(const DNSName& zoneName, std::shared_ptr<CatalogZone>& oldZone, uint32_t& refresh, uint64_t configGeneration, ZoneWaiter& waiter, Logr::log_t logger);
+  bool zoneTrackerIteration(const DNSName& zoneName, std::shared_ptr<CatalogZone>& oldZone, uint32_t& refresh, bool& skipRefreshDelay, uint64_t configGeneration, ZoneWaiter& waiter, Logr::log_t logger);
```
Comment (rgacogne) [inline pdns/recursordist/rec-xfr.hh:144]:
```suggestion
  void preloadZoneFile(const DNSName& zoneName, const std::shared_ptr<const CatalogZone>& oldZone, uint32_t& refresh, uint64_t configGeneration, ZoneWaiter& waiter, Logr::log_t logger);
  bool zoneTrackerIteration(const DNSName& zoneName, std::shared_ptr<const CatalogZone>& oldZone, uint32_t& refresh, bool& skipRefreshDelay, uint64_t configGeneration, ZoneWaiter& waiter, Logr::log_t logger);
```

File: pdns/recursordist/rec-xfr.cc [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+  }
+  if (!zone->dupsCheck()) {
+    zone->clear();
+    throw PDNSException("duplicate PTR values in catalog zone");
+  }
+  logger->info(Logr::Info, "Zone load completed", "nrecords", Logging::Loggable(nrecords), "soa", Logging::Loggable(soaRecordContent->getZoneRepresentation()));
+  return soaRecordContent;
+}
+
+void FWCatZoneXFR::preloadZoneFile(const DNSName& zoneName, std::shared_ptr<CatalogZone>& oldZone, uint32_t& refresh, uint64_t configGeneration, ZoneWaiter& waiter, Logr::log_t logger)
```
Comment (rgacogne) [inline pdns/recursordist/rec-xfr.cc:234]:
```suggestion
void FWCatZoneXFR::preloadZoneFile(const DNSName& zoneName, const std::shared_ptr<const CatalogZone>& oldZone, uint32_t& refresh, uint64_t configGeneration, ZoneWaiter& waiter, Logr::log_t logger)
```

File: pdns/recursordist/rec-xfr.cc [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+    if (luaconfsLocal->generation != configGeneration) {
+      /* the configuration has been reloaded, meaning that a new thread
+         has been started to handle that zone and we are now obsolete.
+      */
+      return;
+    }
+  }
+}
+
+bool FWCatZoneXFR::zoneTrackerIteration(const DNSName& zoneName, std::shared_ptr<CatalogZone>& oldZone, uint32_t& refresh, bool& skipRefreshDelay, uint64_t configGeneration, ZoneWaiter& waiter, Logr::log_t logger)
```
Comment (rgacogne) [inline pdns/recursordist/rec-xfr.cc:286]:
```suggestion
bool FWCatZoneXFR::zoneTrackerIteration(const DNSName& zoneName, std::shared_ptr<const CatalogZone>& oldZone, uint32_t& refresh, bool& skipRefreshDelay, uint64_t configGeneration, ZoneWaiter& waiter, Logr::log_t logger)
```

File: pdns/recursordist/rec-xfr.cc [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+// coverity[pass_by_value] params is intended to be a copy, as this is the main function of a thread
+void FWCatZoneXFR::zoneXFRTracker(ZoneXFRParams params, uint64_t configGeneration) // NOLINT(performance-unnecessary-value-param)
+{
+  setThreadName("rec/fwcatzixfr");
+  bool isPreloaded = params.soaRecordContent != nullptr;
+  auto logger = g_slog->withName("fwcatzixfr");
+  ZoneWaiter waiter(std::this_thread::get_id());
+
+  /* we can _never_ modify this zone directly, we need to do a full copy then replace the existing zone */
+  std::shared_ptr<CatalogZone> oldZone;
```
Comment (rgacogne) [inline pdns/recursordist/rec-xfr.cc:469]:
```suggestion
  std::shared_ptr<const CatalogZone> oldZone;
```

**Review by rgacogne — CHANGES_REQUESTED** [review]
I made a few comments but the code looks generally good and I love the feature, nice work!

File: pdns/recursordist/rec-xfr.cc (hunk 3 above)
Comment (omoerbeek) [reply pdns/recursordist/rec-xfr.cc:174]: The class started with the goal to use a single (base or template) class for both RPZ and catz transfers, but it turned out the goal was a bit ambitious. Some day I might unify the common code. What you are seeing is remains of the original goal. 

The subsystem used with the log messages is "fwcatzixfr", so even though the log message is generic, the subsystem value makes the purpose clear I think.

File: pdns/recursordist/settings/rust/src/bridge.rs [hunk 8]
```diff
@@ -634,6 +655,22 @@ impl TSIGTriplet {
     }
 }
 
+impl ForwardingCatalogZone {
+    pub fn validate(&self, _field: &str) -> Result<(), ValidationError> {
+        Ok(())
+    }
+
+    fn to_yaml_map(&self) -> serde_yaml::Value {
+        // XXX INCOMPLETE
```
[the remark this answers is not in the document]
Comment (omoerbeek) [reply pdns/recursordist/settings/rust/src/bridge.rs:664]: yes, will fix

File: pdns/recursordist/rec-xfr.hh [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+#include "lock.hh"
+#include "logr.hh"
+#include "dnsrecords.hh"
+#include "rust/lib.rs.h"
+
+class DNSName;
+class SOARecordContent;
+struct FWCatz;
+
+// Please make sure that the struct below only contains value types since they are used as parameters in a thread ct
```
Comment (kpfleming) [inline pdns/recursordist/rec-xfr.hh:41]:
```suggestion
// All members of this struct must be copyable, as they are used as parameters in a thread constructor
```
I think the important attribute here is that the members of the struct be *copyable*, not that they are *value* types (in fact `std::shared_ptr` is sometimes considered a reference type). As long as all the members are safe to copy, that's sufficient.

File: pdns/recursordist/rec-xfr.hh [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+{
+  ZoneXFRParams d_params;
+  std::map<std::string, pdns::rust::settings::rec::FCZDefault> d_defaults;
+  std::shared_ptr<CatalogZone> d_catz;
+};
+
+class ZoneXFR
+{
+public:
+  // A struct that holds the condition var and related stuff to allow notifies to be sent to the tread owning
```
Comment (kpfleming) [inline pdns/recursordist/rec-xfr.hh:106]:
```suggestion
  // A struct that holds the condition var and related stuff to allow notifies to be sent to the thread owning
```

File: pdns/recursordist/rec-xfr.hh [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+#include "iputils.hh"
+#include "lock.hh"
+#include "logr.hh"
+#include "dnsrecords.hh"
+#include "rust/lib.rs.h"
+
+class DNSName;
+class SOARecordContent;
+struct FWCatz;
```
Comment (kpfleming) [inline pdns/recursordist/rec-xfr.hh:39]: This is really a nit, but everywhere else the 'zone' part of 'catalog zone' is capitalized in the name... except this one. Should this be `FWCatZ` instead?

File: pdns/recursordist/rec-xfr.cc (hunk 3 above)
Comment (rgacogne) [reply pdns/recursordist/rec-xfr.cc:174]: Fine by me, then :)

File: pdns/recursordist/settings/docs-new-preamble-in.rst [hunk 12]
```diff
@@ -472,6 +472,57 @@ An example of an ``proxymappings`` entry, which is a sequence of `ProxyMapping`_
...
+       tsig:
+         name: string
+         algo: string
+         secret: base64string
+       refresh: number, default not set
+       maxReceivedMBytes: number, default not set
+       localAddress: IP address, default not set
+       axfrTimeout: number, default 20
+     groups:
+     - name:
```
Comment (kpfleming) [inline pdns/recursordist/settings/docs-new-preamble-in.rst:495]:
```suggestion
     - name: optional group name
```

**Review by kpfleming — COMMENTED** [review]
A few comments, nothing serious.

**Review by rgacogne — APPROVED** [review]
:+1:

**Review by chbruyand — APPROVED** [review]

**Pull request merged**